File: robot-impl/src/main/java/ru/avca/robot/factory/GrpcFactory.java
package ru.avca.robot.factory;

import io.grpc.ManagedChannel;
import io.micronaut.context.annotation.Bean;
import io.micronaut.context.annotation.Factory;
import io.micronaut.grpc.annotation.GrpcChannel;
import ru.avca.grpcservices.RobotStateServiceGrpc;
import ru.avca.grpcservices.TradeNotifierGrpc;

import javax.inject.Named;

/**
 * @author a.chermashentsev
 * Date: 19.04.2021
 **/
@Factory
public class GrpcFactory {
    @Named("tg-bot-notifier")
    @Bean
    TradeNotifierGrpc.TradeNotifierFutureStub tgBotFutureStub(
            @GrpcChannel("tg-bot-notifier") ManagedChannel channel
    ) {
        return TradeNotifierGrpc.newFutureStub(channel);
    }

    @Named("database-persist")
    @Bean
    TradeNotifierGrpc.TradeNotifierFutureStub dbPersistFutureStub(
            @GrpcChannel("database-persist") ManagedChannel channel
    ) {
        return TradeNotifierGrpc.newFutureStub(channel);
    }

    @Bean RobotStateServiceGrpc.RobotStateServiceBlockingStub robotStateServiceBlockingStub(
            @GrpcChannel("database-persist") ManagedChannel channel
    ) {
        return RobotStateServiceGrpc.newBlockingStub(channel);
    }
}

File: tg-bot-notifier/src/main/kotlin/ru/avca/grpc/GrpcFactory.kt
package ru.avca.grpc

import io.grpc.ManagedChannel
import io.micronaut.context.annotation.Bean
import io.micronaut.context.annotation.Factory
import io.micronaut.grpc.annotation.GrpcChannel
import ru.avca.grpcservices.RobotStateManagerGrpc

/**
 *
 * @author a.chermashentsev
 * Date: 18.05.2021
 **/
@Factory
open class GrpcFactory {

    @Bean
    open fun robotImpl(
        @GrpcChannel("robot-impl") channel: ManagedChannel
    ): RobotStateManagerGrpc.RobotStateManagerBlockingStub {
        return RobotStateManagerGrpc.newBlockingStub(channel);
    }
}